### PR TITLE
Implement kspace_modify collective yes in KOKKOS package

### LINKED ISSUE #4133: [BUG] Strange communication lapse while scaling PPPM to extreme scale
Hi all

I have been scaling the PPPM long-range potential out to Frontier scale with an SPC/E water box (targeting 9261 nodes). I am using Kokkos+HIP backend. I noticed that I got *very* good weak-scaling behavior, much better than expected, but with an odd "shoulder" in the scaling -- using 64 nodes as the base of the weak-scaling, 512 nodes had much worse scaling than 4096 nodes did (parallel efficiency went down, then back up). I attached the plot of this to this ticket. I found this strange, so I poked further.

I analyzed the bytes sent to/from each MPI rank during a 50-timestep run using an MPI tracing tool I built with LLNL's `wrap` utility. I simply dumped message size and destination for every MPI call. MPI_Send is the dominant MPI call that sends the majority of particle and PPPM data, so I focused on this.

I chose the rank with the maximum time in MPI_Send and computed the total MPI_Send's called and total number of bytes for that rank for each node count, here is what I found:

Nodecount: max MPI_Send time, total bytes, total messages
64-node: 64 seconds, 453.1x10^9 bytes, 41.2K messages
512-node: 115 seconds 435.2x10^9 bytes, 251.3K messages
4096-node: 89 seconds, 481.5x10^9 bytes, 98.8K messages
9261-node: 188 seconds, 473.9x10^9 bytes, 136.5K messages

The 512-node point didn't look right, so next, I looked at *where* the data was sent. I've attached 4 plots to this ticket named "r2a_n<nodecount>_rank1_count.png". This shows the number of MPI_Send's called by rank 1 to each of the destination ranks on the x-axis. From what I know of LAMMPS, there's 2 main reasons to communicate with MPI_Send -- particle/neighbor list exchange and PPPM FFT data exchange. I'd expect to see 6 "peak" ranks where particle exchange happens -- one neighbor in each cartesian direction. Then I'd expect every other rank in the simulation to receive data from rank 1 for FFTs. But, looking at the plots, this is not true -- at 64 and 512 nodes, it is true, but at 4096 nodes, there's a large number of ranks not communicated with by rank 1. Do you have any insight? I am happy to help put time towards figuring this out, I'm just not sure where to look.

I also ran this problem on Summit, but 4096 Summit nodes can only run the 512-Frontier-node problem due to HBM sizes. Summit showed similar scaling behavior up to the 512-node problem (run on 4096 nodes of Summit).

**LAMMPS Version and Platform**

LAMMPS commit 08498637aa built for Kokkos+HIP with hipcc from ROCm/5.7.1 + cray-mpich/8.1.28, uses hipFFT.
Run on Frontier, but also seems to be trending in this direction on Summit.

**Expected Behavior**

I'd expect the message count for MPI_Send to scale consistently, not shoot up exponentially, then drop back down at a large enough node count.

**Actual Behavior**

Rank 1 calls MPI_Send 41.2K, 251.3K, 98.8K, and 136.5K times at 64, 512, 4096, and 9261 nodes respectively. The 4096+ node points seem wrong.

**Steps to Reproduce**

I can provide input files as needed. Using an srun line like:
```
srun -N ${SLURM_NNODES} -n $(expr ${SLURM_NNODES} \* 8) -c 7 --gpus-per-task=1 --gpu-bind=closest \
    lmp -k on g 1 -sf kk -pk kokkos gpu/aware on neigh half comm device newton off neigh/transpose on sort device atom/map no \
    -in input.in -log none -v x ${x} -v y ${y} -v z ${z}
```

**Further Information, Files, and Links**

64-node:

r2a_n64_rank1_count

512-node:

r2a_n512_rank1_count

4096-node:

r2a_n4096_rank1_count

9261-node:

r2a_n9261_rank1_count

Parallel efficiency plot:

lammps_parallel_efficieny

### LINKED ISSUE #4140: [Feature Request] Implement MPI collective operations in KOKKOS implementation of KSPACE
Creating this issue to track the implementation of `kspace_modify collective yes` in Kokkos-enabled KSPACE package.
This has already been implemented in the non-Kokkos version of KSPACE.

**Detailed Description**

Collective code in the regular (non-Kokkos) node here:
https://github.com/lammps/lammps/blob/develop/src/KSPACE/remap.cpp#L113-L203
which is missing from the Kokkos version: https://github.com/lammps/lammps/blob/develop/src/KOKKOS/remap_kokkos.cpp#L106

The Kokkos version needs to support both GPU-aware MPI on and off, which is already done for the point-to-point version.

**Further Information, Files, and Links**

### PR DESCRIPTION
**Pull request by hagertnl**
**Summary**

Implements the `kspace_modify collective yes` option in the KOKKOS package with KSPACE enabled.

**Related Issue(s)**

closes #4140
closes #4133 

**Author(s)**

Nick Hagerty, Oak Ridge National Laboratory

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No change in backwards compatibility

**Implementation Notes**

A simple water box with PPPM enabled was run on 1 and 8 ranks on a single node, and on 512 ranks, 4096 ranks, 32K ranks for scaling analysis.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by stanmoore1** [comment]
>A simple water box with PPPM enabled was run on 1 and 8 ranks on a single node, and on 512 ranks, 4096 ranks, 32K ranks for scaling analysis.

@hagertnl this wasn't actually being used in `PPPMKokkos`, may need to rerun the tests

**Comment by stanmoore1** [comment]
I tested this on 40 MPI ranks on CPU and on 4 V100 GPUs and it seems to give the correct numerics, nice work @hagertnl. Really curious what your scaling analysis will show, especially at large-scale.

**Comment by hagertnl** [comment]
I figured that out when my performance numbers were turning out to be the exact same, hah! I still need to tie in the `kspace_modify` keyword to control the setting (right now I just locally hard-coded in a 1 for collectives).

Preliminary, at 512 nodes (4096 ranks), I see about a 13% speedup, with correct answers. I'd like to do a quick round of optimization, I think the setup is doing double the work that it needs to for the collectives, creating excessive overhead for the collective approach.

Not surprisingly, small scale did not see any benefit, and saw some performance regression in the case I saw. This is why I wanted to do a round of optimization, it seems like there may be excessive overhead.

**Comment by hagertnl** [comment]
Here are some preliminary numbers from Frontier, 8 ranks per node (MPI timings from rank 1):

| # Nodes | coll/pt2pt | timesteps/s | Total MPI time | MPI_Send time | MPI_Alltoallv time |
|----------|----------|-------------|----------------|---------------|--------------------|
| 64           | coll          | 0.316            | 26.91                 | 4.06                 | 19.40                        |
| 64           | pt2pt      | 0.360            | 65.15                 | 58.30               | 0.17                          |
| 512         | coll         | 0.288             | 52.30                | 3.76                  | 37.02                        |
| 512         | pt2pt      | 0.258             | 121.64               | 111.13               | 0.53                          |
| 4096      | coll         | 0.161              | 119.88               | 2.92                 | 76.71                         |
| 4096      | pt2pt      | 0.300             | 112.38               | 76.03               | 16.56                        |

Some of these jobs hit network timeouts, and these timings aren't averaged, so these numbers may change slightly run-to-run. These runs are also without the patch from #4133 , so attempting to draw conclusions across job sizes is tricky.

The conclusions I draw from this currently are:
- At 64 and 512 nodes, there's a >50% drop in total MPI time when using collectives relative to pt2pt, but the collective is actually slower per-timestep. I suspect there's a lot of overhead the way I'm currently packing the buffer for Alltoallv
- In spite of the possible overhead, we still see a >10% speedup at 512 nodes.

@stanmoore1 currently, the collective code for kokkos and non-kokkos PPPM are fairly apples-to-apples. From your end, would you prefer any of the following, with respect to the scope of work in this PR? (1) finish verifying the correctness of the implementation we have now, then merge, optimizing both kokkos and non-kokkos PPPM collectives in a new PR later, (2) optimize just kokkos PPPM now (diverging the Kokkos implementation of remap from non-Kokkos), (3) optimize kokkos & non-kokkos PPPM collectives now, or something else?

**Comment by stanmoore1** [comment]
@hagertnl it is odd that the 512 node case is slower than 4096. Do these results include the patch with better load balancing by not using full xy planes, like we saw previously?

>From your end, would you prefer any of the following, with respect to the scope of work in this PR? (1) finish verifying the correctness of the implementation we have now, then merge, optimizing both kokkos and non-kokkos PPPM collectives in a new PR later, (2) optimize just kokkos PPPM now (diverging the Kokkos implementation of remap from non-Kokkos), (3) optimize kokkos & non-kokkos PPPM collectives now, or something else?

This is mostly up to you and how much time you have to work on this. If you have more time it would be great to optimize more, otherwise we can merge now as-is.

**Comment by hagertnl** [comment]
This does not include the patch for better load balancing. Is that something you'd like me to put in this PR? I thought it might've been in another PR already.

**Comment by stanmoore1** [comment]
I think the load balancing issue is confounding the `pt2pt` vs `coll` comparison in the table above. It sounded like Steve agreed that we should apply that patch since we aren't using 2D FFTs when we have full xy planes. It is not in a PR yet.

**Comment by hagertnl** [comment]
Sounds good. I'll apply the patch in this PR and try out some optimization.

**Comment by hagertnl** [comment]
This is close to what I hope is the final version. I am currently running scaling studies on Frontier & Summit to test. At small scale, pt2pt is 10%+ faster than collective, but that's to be expected. I'm very interested what happens at 512+ nodes.

Note -- I replaced a ~100-line section of code that looked for what other ranks needed to communicate for commringlist with a single Allreduce. I assume that if the user wants to use Alltoallv, then the Allreduce should be performant on the platform as well, and it saves running a nested for-loop that gets very expensive at scale. I did some testing of this vs the old 100-line section of code and it is performance neutral for 1-64 nodes. I did not have a chance to run at 512+, but unless the Allreduce is prohibitively expensive, I don't see a reason that the local commringlist building would be faster or preferred.

**Comment by stanmoore1** [comment]
>I assume that if the user wants to use Alltoallv, then the Allreduce should be performant on the platform as well, and it saves running a nested for-loop that gets very expensive at scale. 

I agree with this.

**Comment by hagertnl** [comment]
In profiling (to validate message sizes), I noticed that ranks send themselves a significant amount of data. For example, in the below Alltoallv call, there were about 746 MB sent total by the MPI_Alltoallv, and half of that -- 373 MB -- was sent to the current process:
```
[Rank 0] MPI_Alltoallv started 1714096596.363301992, ended 1714096596.480638027 (elapsed 0.117336035), sent 746496000 bytes total, max of 373248000 to rank 0, received 746496000 bytes total, max of 373248000 from rank 0, to MPI_COMM_WORLD
```
I am thinking it is worth checking the performance impact of implementing `plan->self` like non-collective strategy does, since it can substantially decrease the total bytes sent by MPI. Ideally, MPI implementations are smart enough to use a local copy for Alltoallv messages to self, but that doesn't appear to be true with cray-mpich.

**Comment by stanmoore1** [comment]
>I am thinking it is worth checking the performance impact of implementing plan->self like non-collective strategy does, since it can substantially decrease the total bytes sent by MPI. Ideally, MPI implementations are smart enough to use a local copy for Alltoallv messages to self, but that doesn't appear to be true with cray-mpich.

Interesting, yes that would be good to check.

**Comment by hagertnl** [comment]
Just a status update, currently I'm debugging an issue on Infiniband networks, where the NIC appears to be reporting that it's out of physical memory while doing the Alltoallv. Once I have that worked out, I'll do a quick scaling study with small & large system sizes on both Summit and Frontier.

For separating communication to self out from MPI, I observed a small performance improvement at small node counts (still slower than pt2pt), but no observable improvement at large scale on Frontier. I was trying to test this on Summit when I hit the Infiniband error. I'll test both implementations on Frontier & Summit when I do the scaling study, and we can figure out which implementation we want based on the data.

**Comment by stanmoore1** [comment]
@hagertnl very interesting, thanks for working on this.

**Comment by hagertnl** [comment]
Performance data from Frontier (Slingshot 11, Dragonfly topology) and Summit (Infiniband, Fat-tree topology) for small & large systems. Summit's "large" problem is 8x smaller per-node than Frontier's, due to HBM limits. The small problems are 8x smaller per-node than large for each system. In all cases, I ran 3 jobs and took the average of timesteps/sec, excluding obvious outliers.

Regarding the issues at 4096 nodes I saw on Summit, it appears to be due to this: https://www.open-mpi.org/faq/?category=openfabrics#ib-locked-pages. Basically, we're just overrunning the resources available. There are some parameters we can tune to get it to work, but I don't know them off the top of my head. Smaller systems were able to run with collectives without a problem.

Screenshot 2024-05-11 at 8 54 19 PM

**Comment by hagertnl** [comment]
Just pushed the source for the collective+self optimization. I still need to clean up the memory deallocations. The remap destructor is not organized like the allocations, and I think I'm probably allocating more arrays than necessary, or I can re-use some of the existing pt2pt variables.

**Comment by stanmoore1** [comment]
@hagertnl looks like the collective option can give significant speedup on Frontier, which is great.

**Comment by hagertnl** [comment]
@stanmoore1 it's actually opposite (metric reported was timesteps/sec) -- it's quite a slow-down on Frontier at scale. But it is quite an improvement on full-system Summit (>2x faster for small systems).

Frontier has a much higher message injection rate into the fabric due to the 4 NICs per node (only 1 NIC per node on Summit), so I'm not very surprised by these results. Also, cray-mpich is still being tuned for AMD GPU support, so some lesser-used collectives may not be fully optimized.

Also note that Summit's pt2pt results are tuned with Infiniband network settings that give a ~50% speedup from default settings, so it's impressive that collectives still beat that.

All that to say, I think that the benefit of these optimizations depends more heavily on the node design than I originally thought. If the NIC's injection rates are not limiting the MPI_Send's, then it's entirely possible that collectives won't be optimal. One interesting thing to try could be to implement MPI_Isend's instead of Sends by using the full send buffer as we do with collectives so that each rank has an independent portion of the buffer, but that's an addition for later, I think.

**Comment by stanmoore1** [comment]
>@stanmoore1 it's actually opposite (metric reported was timesteps/sec) -- it's quite a slow-down on Frontier at scale. But it is quite an improvement on full-system Summit (>2x faster for small systems).

Ah got it, interesting...

**Comment by hagertnl** [comment]
I believe this code is ready to go then -- I looked at the docs to see if they need updates, but it doesn't look like it.

What I currently have pushed is the implementation where the current processor's data is locally copied and is not sent through MPI. This was faster or comparable in all cases above, and is especially beneficial for small scale.

**Comment by stanmoore1** [comment]
@hagertnl thanks for all your effort on this. How beneficial do you think it would be to port the changes to Remap to non-Kokkos? (Doesn't need to be done by you or in this PR, just curious...).

**Comment by hagertnl** [comment]
That's a good question... The main performance changes I could see is in that nested for-loop that locally builds the commringlist that I replaced with MPI_Allreduce, and it's hard to say -- what will be faster, an Allreduce with a lot of ranks, or a double-nested for-loop for a lot of iterations in each? I think as number of ranks goes up, the Allreduce *should* be faster. And since most processes send a lot of data to themselves, pulling that communication out of the Alltoallv into a local pack/unpack could be beneficial at all scales.

It also would contribute to code maintainability, I think. There's a few spots that I felt like the logic was tricky to follow, I think the Kokkos implementation is more readable than the non-Kokkos version was. Mainly just getting larger conditional blocks for `if (!plan->use_collective) { ... } else { ... }`. When the conditional blocks get interleaved, it was tricky to follow.

Given that there's an existing implementation, it would be neat to see if there's any performance difference from the changes!

**Comment by stanmoore1** [comment]
@hagertnl I tried this on V100 with the standard `lammps/bench/in.rhodo` but I'm seeing a segfault:

```
PPPM Kokkos initialization ...
  using 12-bit tables for long-range coulomb (../kspace.cpp:342)

Program received signal SIGSEGV, Segmentation fault.
0x000000001172f954 in LAMMPS_NS::RemapKokkos<Kokkos::Cuda>::remap_3d_create_plan_kokkos (this=<optimized out>, comm=0x7ffff67612a0 <ompi_mpi_comm_world>, 
    in_ilo=<optimized out>, in_ihi=<optimized out>, in_jlo=<optimized out>, 
    in_jhi=<optimized out>, in_klo=<optimized out>, in_khi=<optimized out>, 
    out_ilo=0, out_ihi=31, out_jlo=0, out_jhi=24, out_klo=0, out_khi=31, 
    nqty=2, permute=1, memory=0, usecollective=1, usegpu_aware=1)
    at ../remap_kokkos.cpp:273
273	struct remap_plan_3d_kokkos<DeviceType>* RemapKokkos<DeviceType>::remap_3d_create_plan_kokkos(
Missing separate debuginfos, use: debuginfo-install glibc-2.17-324.el7_9.ppc64le libgcc-4.8.5-44.el7.ppc64le libibverbs-41mlnx1-OFED.4.7.0.0.2.47329.ppc64le libmlx4-41mlnx1-OFED.4.7.3.0.3.47329.ppc64le libmlx5-41mlnx1-OFED.4.7.0.3.3.47329.ppc64le libnl3-3.2.28-4.el7.ppc64le librxe-41mlnx1-OFED.4.4.2.4.6.47329.ppc64le libstdc++-4.8.5-44.el7.ppc64le numactl-libs-2.0.12-5.el7.ppc64le nvidia-driver-cuda-libs-510.47.03-1.el8.ppc64le zlib-1.2.7-19.el7_9.ppc64le
(gdb) where
#0  0x000000001172f954 in LAMMPS_NS::RemapKokkos<Kokkos::Cuda>::remap_3d_create_plan_kokkos (this=<optimized out>, comm=0x7ffff67612a0 <ompi_mpi_comm_world>, 
    in_ilo=<optimized out>, in_ihi=<optimized out>, in_jlo=<optimized out>, 
    in_jhi=<optimized out>, in_klo=<optimized out>, in_khi=<optimized out>, 
    out_ilo=0, out_ihi=31, out_jlo=0, out_jhi=24, out_klo=0, out_khi=31, 
    nqty=2, permute=1, memory=0, usecollective=1, usegpu_aware=1)
    at ../remap_kokkos.cpp:273
#1  0x00000000118dd67c in LAMMPS_NS::FFT3dKokkos<Kokkos::Cuda>::fft_3d_create_plan_kokkos (this=0x24342b50, comm=<optimized out>, nfast=<optimized out>, 
    nmid=<optimized out>, nslow=<optimized out>, in_ilo=<optimized out>, 
    in_ihi=<optimized out>, in_jlo=<optimized out>, in_jhi=31, in_klo=0, 
    in_khi=31, out_ilo=0, out_ihi=24, out_jlo=0, out_jhi=31, out_klo=0, 
    out_khi=31, scaled=0, permute=0, nbuf=0x7fffffff8d50, usecollective=1, 
    nthreads=1, usegpu_aware=1) at ../fft3d_kokkos.cpp:383
#2  0x00000000118ddabc in LAMMPS_NS::FFT3dKokkos<Kokkos::Cuda>::FFT3dKokkos (
    this=<optimized out>, 
    lmp=<error reading variable: value has been optimized out>, 
    comm=<optimized out>, nfast=<optimized out>, nmid=<optimized out>, 
    nslow=<optimized out>, in_ilo=<optimized out>, in_ihi=<optimized out>, 
    in_jlo=0, in_jhi=31, in_klo=0, in_khi=31, out_ilo=0, out_ihi=24, 
    out_jlo=0, out_jhi=31, out_klo=0, out_khi=31, scaled=0, permute=0, 
    nbuf=0x7fffffff8d50, usecollective=1, usegpu_aware=1)
    at ../fft3d_kokkos.cpp:32
#3  0x0000000010d2c5fc in LAMMPS_NS::PPPMKokkos<Kokkos::Cuda>::allocate (
    this=0x2430eb70)
    at ../../lib/kokkos/core/src/impl/Kokkos_SharedAlloc.hpp:617
#4  0x0000000010d2216c in LAMMPS_NS::PPPMKokkos<Kokkos::Cuda>::init (
    this=0x2430eb70) at ../pppm_kokkos.cpp:139
#5  0x00000000104d82d8 in LAMMPS_NS::Force::init (this=0x242e6380)
    at ../force.cpp:186
#6  0x000000001039ff14 in LAMMPS_NS::LAMMPS::init (this=0x191b4490)
    at ../lammps.cpp:955
#7  0x0000000010207394 in LAMMPS_NS::Run::command (this=0x24da4220, 
    narg=<optimized out>, arg=0x242fe550) at ../run.cpp:170
#8  0x00000000100b1060 in LAMMPS_NS::Input::execute_command (this=0x2428c490)
    at ../input.cpp:868
#9  0x00000000100b1ef0 in LAMMPS_NS::Input::file (this=0x2428c490)
    at ../input.cpp:313
#10 0x0000000010021340 in main (argc=<optimized out>, argv=<optimized out>)
    at ../main.cpp:77
```
And here is Valgrind:
```
PPPM Kokkos initialization ...
  using 12-bit tables for long-range coulomb (../kspace.cpp:342)
==10741== Conditional jump or move depends on uninitialised value(s)
==10741==    at 0x1172FEC4: LAMMPS_NS::RemapKokkos<Kokkos::Cuda>::remap_3d_create_plan_kokkos(ompi_communicator_t*, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int) (remap_kokkos.cpp:616)
==10741==    by 0x118DCFBF: LAMMPS_NS::FFT3dKokkos<Kokkos::Cuda>::fft_3d_create_plan_kokkos(ompi_communicator_t*, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int*, int, int, int) (fft3d_kokkos.cpp:466)
==10741==    by 0x118DDABB: LAMMPS_NS::FFT3dKokkos<Kokkos::Cuda>::FFT3dKokkos(LAMMPS_NS::LAMMPS*, ompi_communicator_t*, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int*, int, int) (fft3d_kokkos.cpp:74)
==10741==    by 0x10D2C5FB: LAMMPS_NS::PPPMKokkos<Kokkos::Cuda>::allocate() (pppm_kokkos.cpp:801)
==10741==    by 0x10D2216B: LAMMPS_NS::PPPMKokkos<Kokkos::Cuda>::init() (pppm_kokkos.cpp:263)
==10741==    by 0x104D82D7: LAMMPS_NS::Force::init() (force.cpp:186)
==10741==    by 0x1039FF13: LAMMPS_NS::LAMMPS::init() (lammps.cpp:955)
==10741==    by 0x10207393: LAMMPS_NS::Run::command(int, char**) (run.cpp:170)
==10741==    by 0x100B105F: LAMMPS_NS::Input::execute_command() (input.cpp:868)
==10741==    by 0x100B1EEF: LAMMPS_NS::Input::file() (input.cpp:313)
==10741==    by 0x1002133F: main (main.cpp:77)
==10741== 
==10741== Conditional jump or move depends on uninitialised value(s)
==10741==    at 0x1172FF50: LAMMPS_NS::RemapKokkos<Kokkos::Cuda>::remap_3d_create_plan_kokkos(ompi_communicator_t*, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int) (remap_kokkos.cpp:653)
==10741==    by 0x118DCFBF: LAMMPS_NS::FFT3dKokkos<Kokkos::Cuda>::fft_3d_create_plan_kokkos(ompi_communicator_t*, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int*, int, int, int) (fft3d_kokkos.cpp:466)
==10741==    by 0x118DDABB: LAMMPS_NS::FFT3dKokkos<Kokkos::Cuda>::FFT3dKokkos(LAMMPS_NS::LAMMPS*, ompi_communicator_t*, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int, int*, int, int) (fft3d_kokkos.cpp:74)
==10741==    by 0x10D2C5FB: LAMMPS_NS::PPPMKokkos<Kokkos::Cuda>::allocate() (pppm_kokkos.cpp:801)
==10741==    by 0x10D2216B: LAMMPS_NS::PPPMKokkos<Kokkos::Cuda>::init() (pppm_kokkos.cpp:263)
==10741==    by 0x104D82D7: LAMMPS_NS::Force::init() (force.cpp:186)
==10741==    by 0x1039FF13: LAMMPS_NS::LAMMPS::init() (lammps.cpp:955)
==10741==    by 0x10207393: LAMMPS_NS::Run::command(int, char**) (run.cpp:170)
==10741==    by 0x100B105F: LAMMPS_NS::Input::execute_command() (input.cpp:868)
==10741==    by 0x100B1EEF: LAMMPS_NS::Input::file() (input.cpp:313)
==10741==    by 0x1002133F: main (main.cpp:77)
```

**Comment by hagertnl** [comment]
Hi Stan,
Sorry for the delay -- do you remember how many ranks you used to reproduce this seg fault?

**Comment by stanmoore1** [comment]
>do you remember how many ranks you used to reproduce this seg fault?

I think just one (single GPU).

**Comment by hagertnl** [comment]
Hm. I was able to successfully run the rhodo benchmark with:
```
$ jsrun --nrs 1 --gpu_per_rs 1 --cpu_per_rs 7 --tasks_per_rs 1 --rs_per_host 1 ../install-gpu/bin/lmp -in in.rhodo -sf kk -k on g 1 -pk kokkos gpu/aware off neigh half
```
This is on Summit (LSF) with 1 V100. I also tried with gpu/aware on, and it still worked there. The benchmark fails if I leave out `neigh half`.

**Comment by stanmoore1** [comment]
Let me try again and see if I can reproduce.

**Comment by stanmoore1** [comment]
@hagertnl I still see the segmentation fault. I'm running on SNL Vortex, which is a small testbed for LLNL Sierra. 

Here is my run command: 

`jsrun -M "-gpu -disable_gdr" -r 1 -a 1 -c 1 -g 1 -l gpu-cpu -d packed /lammps/src/lmp_vortex_kokkos -in in.rhodo -k on g 1 -sf kk -pk kokkos neigh half`

The only change I made to the Rhodo input is adding `collective yes`:

```patch
diff --git a/bench/in.rhodo b/bench/in.rhodo
index 2f34b2721b..9a7fdfddad 100644
--- a/bench/in.rhodo
+++ b/bench/in.rhodo
@@ -11,6 +11,7 @@ improper_style  harmonic
 pair_style      lj/charmm/coul/long 8.0 10.0
 pair_modify     mix arithmetic
 kspace_style    pppm 1e-4
+kspace_modify   collective yes
 
 read_data       data.rhodo
 ```
 
 Edit: I'm using `cuda/11.1.1` and `gcc/8.3.1` to compile LAMMPS, along with a GNU Makefile.

**Comment by hagertnl** [comment]
Ah... It would've been helpful if I had turned on the collectives with `collective yes`. I have successfully proven that the default pt2pt version works :)

**Comment by hagertnl** [comment]
I'm not entirely sure how the line that got there got to where it is, but I found the source of the seg-fault. I'm looking through git logs to see where it might've been added.

**Comment by hagertnl** [comment]
@stanmoore1  can you try this? There were some unused buffers from the pt2pt implementation that I somehow looked past when I wrote the initial implementation. Removing those allows me to successfully run the Rhodo benchmark.

**Comment by stanmoore1** [comment]
I ran Rhodo on 1 and 4 GPUs, everything seems to be working correctly.

**Review by stanmoore1 — APPROVED** [review]

**Comment by stanmoore1** [comment]
@akohlmey once the dust settles from the new stable, this PR is ready for review/merging.

**Comment by akohlmey** [comment]
@stanmoore1 @hagertnl This change should be reflected in the documentation. Am I correct that this is for KOKKOS only? The the documentation should emphasize that.

**Comment by stanmoore1** [comment]
There is a CPU version too. I don't think it needs a documentation update. It was a missing Kokkos feature but the docs didn't mention it was missing. Now it exists. @sjplimp should review the change to `pppm.cpp`.

File: src/KSPACE/pppm.cpp
Comment (akohlmey) [inline src/KSPACE/pppm.cpp:1]: @sjplimp please let us know if you are ok with this change.

Comment (stanmoore1) [reply src/KSPACE/pppm.cpp:1]: Steve told me he approves, but will probably officially click the button soon...

**Review by sjplimp — APPROVED** [review]
